src/lavinmq/auth/authenticators/oauth.cr

```diff
@@ -0,0 +1,410 @@
+require "../authenticator"
+require "../users/oauth_user"
+require "../../config"
+require "../jwt"
+require "openssl"
+require "http/client"
+require "../lib_crypto_ext"
+
+module LavinMQ
+  module Auth
+    class OAuthAuthenticator < Authenticator
+      Log = LavinMQ::Log.for "oauth2"
+
+      @cached_public_keys : Hash(String, String)?
+      @cache_expires_at : Time?
+      @cache_mutex : Mutex = Mutex.new
+
+      # OAuth 2.0 / OpenID Connect authenticator for LavinMQ.
+      #
+      # Clients authenticate by providing a JWT token (RS256-signed) as the password field during
+      # AMQP/MQTT connection. The username field is used for logging only - authorization is
+      # based on permissions/scopes extracted from the JWT token.
+      #
+      # Verification flow:
+      # 1. Prevalidate: Check JWT format, algorithm (RS256), and expiration
+      # 2. Verify signature: Decode JWT using OAuth provider's public keys (from JWKS)
+      # 3. Validate claims: Check issuer and optionally audience match configuration
+      # 4. Extract permissions: Parse scopes/roles and map to LavinMQ permissions
+      # 5. Create OAuthUser: With extracted username, tags, permissions, and expiration time
+      #
+      # Token refresh: Clients send UpdateSecret frame with new JWT to update permissions
+      # without reconnecting. Username stays the same, only permissions/expiration are updated.
+      #
+      # Public Key caching: Public keys are fetched on first use and cached with TTL from the
+      # JWKS endpoints Cache-Control header (or config default). Keys are re-fetched when
+      # cache expires.
+      #
+      # Security: Only RS256 accepted (prevents algorithm confusion), issuer/audience
+      # validation prevents accepting tokens from untrusted or unintended sources, fails
+      # closed on any error.
+      def initialize(@config = Config.instance)
+      end
+
+      def authenticate(username : String, password : Bytes) : OAuthUser?
+        token = String.new(password)
```

Comment:
it's only in the mqtt protocl that the pssword can be any byte array. AMQP requires it to be an UTF8 string, and so does all other authenticators. now, in authenticator it's converted to a slice, and then back to string again here.